compiler/hash-reporting/src/render.rs

```diff
@@ -0,0 +1,464 @@
+
+        bytes_skipped += skip_width;
+        total_lines += 1;
+        last_line_len = line.len();
+    }
+
+    line_index.unwrap_or((
+        last_line_len.saturating_sub(1),
+        total_lines.saturating_sub(1),
+    ))
+}
+
+/// This function holds inner rules for calculating what the selected top
+/// and bottom buffer sizes should be.
+fn adjust_initial_span_size(span: usize) -> usize {
+    // @@Correctness: This is an arbitrary rule and should be decided upon more concretely. It's
+    //                even possible that we might not want to get any additional lines for even larger
+    //                spans
+    match span {
+        3..=4 => 2,
+        5.. => 1,
+        _ => span,
+    }
+}
+
+/// Compute the top buffer and bottom buffer sizes from the given start and
+/// end row for a code block.
+fn compute_buffers(start_row: usize, end_row: usize) -> (usize, usize) {
+    const MAX_BUFFER: usize = 5;
+
+    let diff = adjust_initial_span_size(end_row - start_row + 1);
+
+    // Cap the buffers to a maximum of 5
+    let top_buffer = diff.min(MAX_BUFFER);
+    let bottom_buffer = diff.min(MAX_BUFFER);
+
+    (top_buffer, bottom_buffer)
+}
+
+impl ReportCodeBlock {
+    // Get the indent widths of this code block as (outer, inner).
+    pub(crate) fn info<T: SourceMap>(&self, modules: &T) -> ReportCodeBlockInfo {
+        match self.info.get() {
+            Some(info) => info,
+            None => {
+                let SourceLocation { span, source_id } = self.source_location;
+                let source = modules.contents_by_id(source_id);
+
+                // Compute offset rows and columns from the provided span
+                let (start_col, start_row) = offset_col_row(span.start(), source, true);
+                let (end_col, end_row) = offset_col_row(span.end(), source, false);
+                let (_, last_row) = offset_col_row(source.len(), source, false);
+
+                // Compute the selected span outside of the diagnostic span
+                let (top_buf, bottom_buf) = compute_buffers(start_row, end_row);
+
+                // Compute the size of the indent based on the line numbers
+                let indent_width = (start_row.saturating_sub(top_buf) + 1)
+                    .max((end_row + bottom_buf).min(last_row) + 1)
+                    .to_string()
+                    .chars()
+                    .count();
+
+                let info = ReportCodeBlockInfo {
+                    indent_width,
+                    start_col,
+                    start_row,
+                    end_col,
+                    end_row,
+                };
+
+                self.info.replace(Some(info));
+                info
+            }
+        }
+    }
+
+    /// Function to extract the block of the code that will be used to display the span
+    /// of the diagnostic.
+    fn get_source_view<'a, T: SourceMap>(
+        &self,
+        modules: &'a T,
+    ) -> Take<Skip<Enumerate<Split<'a, char>>>> {
```

Comment:
Is it possible to just write `impl Iterator<Item=&str>` here?

Reply:
apparently you can't

Reply:
Also, it would be `char` not `str` right